specs/merge/beacon-chain.md

```diff
@@ -257,6 +277,9 @@ def process_block(state: BeaconState, block: BeaconBlock) -> None:
     process_eth1_data(state, block.body)
     process_operations(state, block.body)
     process_sync_aggregate(state, block.body.sync_aggregate)
+    if is_execution_enabled(state, block.body):
```

Comment:
This is not technically the appropriate place to run `consensus_validated` because `block.state_root` is not validated yet. Instead, it should be after `if validate_result` in `state_transition()`.

You might debate whether it should be embedded in `if validate_result` code block or outside. `validate_result == False` only when a proposer is building a beacon block (and thus when computing the state root to embed rather validating some untrusted input).

To that end, we should discuss what the execution engine does with an execution_payload that it produced for a local proposer via `engine_getPayload`. Must this be inserted manually via `engine_executePayload` and then we run `engine_consensusValidated`? Or instead does the execution engine keep the work done in `engine_getPayload` then either assuming consensus validated or awaiting consensus validated?

Those answers imply how to actually spec this here, in the validator guide, and in the engine API.

My gut is that the execution engine upon a call `engine_getPayload` operates in a similar way wrt to the block as though `engine_executePayload` has been called, but not yet `engine_consensusValidated`. And that only once the proposer has fully computed and signed/broadcast the beacon block, does the beacon node call `engine_consensusValidated`. So I'd suggest the following changes:

* Move `EXECUTION_ENGINE.consensus_validated(block.body.execution_payload)` in the `state_transition()` inside the `if validate_result:` codeblock
* Add in the validator guide that after the validator has successfully created/signed the block, that they call `EXECUTION_ENGINE.consensus_validated` -- note that I think if `validate_result == False` that it is too early to call it within `state_transition()` because the final signature has not been made on the block thus there is still room for the validator to fail/abort the process.
* Add a note in the engine API on `engine_getPayload` that this should leave the node the same state as though `engine_executePayload` was called -- transition made but awaiting final confirmation from `engine_consensusValidated

Reply:
> This is not technically the appropriate place to run `consensus_validated` because `block.state_root` is not validated yet. Instead, it should be after `if validate_result` in `state_transition()`.

This is so true!

> Add a note in the engine API on `engine_getPayload` that this should leave the node the same state as though `engine_executePayload` was called -- transition made but awaiting final confirmation from `engine_consensusValidated`

I would not mix proposal and processing flows as this would add one more state to the block proposal state machine, i.e. `prepare + get + validated` instead of `prepare + get`. The spec should also be cleaner if we define them separate.

I am not sure if every consensus client does the block proposal as follows but Teku does and it looks like an optimal solution:
1. VC requests unsigned block from BN
2. VC signs it and submits it to the BN
3. BN gossips signed block
4. BN imports signed block as usual

This approach makes it sure that block gets gossiped ASAP and fully verified before attesting to this block happens. If we add execution client to this scheme it will start looking as follows:
1. VC requests unsigned block from BN
    * BN calls `getPayload` and returns a block with the payload
2. VC signs it and submits it to the BN
3. BN gossips signed block
4. BN imports signed block as usual
    * BN calls `executePayload`
    * BN calls `consensusValidated`

In this case execution client will do double work on executing the payload. I don't think it matter much as long as block gossip is not affected by the execution. Execution client may look for an optimisation in a way they store the block hash mapped on the state when `getPayload` is served expecting that this payload will be sent a few moments later with `executePayload` and avoid executing the same payload twice.

Reply:
That looks good to me.

I don't want extra work to be done, but you're right that an optimistic cache of the result of prepare/get can be used to mitigate that without having to rely upon alterations of the API. Additionally, you're right that we should mirror the normal BN block import path as much as possible

Reply:
I'm okay with that for now. Will think about if we want to represent that fully in the `state_transition`. It might be useful for testing.

Reply:
I was also thinking about test scenarios and the value that spec'ing this call inside of the `state_transition` may bring in. Implementation of a function on such a high level may vary from client to client and additional tweaks would be required to adopt these test vectors. For instance, implementations may make this call outside of the `state_transition()` function, after it finishes its work.

By the end of the day, this method call won't stay uncovered in terms of testing as it should be a part of integration tests.

Reply:
> this method call won't stay uncovered in terms of testing as it should be a part of integration tests.

I hear you on that, but in terms of completeness, especially in relation to EIP 3675, the state transition function is not really complete until sending that signal.

I also see the argument for not including it. Can leave as is and assess it's integration as we do more testing.

Reply:
I hear arguments on having this method call as a part of the state transition. What do you think about `notify_forkchoice_updated`? Should we make calls to this function in the fork choice handlers in a similar vein?